### ISSUE TITLE
Composite column is already fully constructed

### ISSUE THREAD

**Issue by rpazuki** [Stratio/cassandra-lucene-index#143]
Hi,

I am running  "Cassandra 3.0.5" with "cassandra-lucene-index-plugin-3.0.5.2-RC1-SNAPSHOT.jar" plugin.

This is my table:

```
CREATE TABLE ds1.waypoints_subset (
    id int,
    year int,
    month int,
    day int,
    hour int,
    minute int,
    second int,
    datetime timestamp,
    driverid int,
    driverid int,
    idx1col text,
    ignition boolean,
    lat double,
    lon double,
    odometer double,
    speed double,
    PRIMARY KEY (id, year, month, day, hour, minute, second)
) WITH CLUSTERING ORDER BY (year ASC, month ASC, day ASC, hour ASC, minute ASC
, second ASC)
    AND bloom_filter_fp_chance = 0.01
    AND caching = {'keys': 'ALL', 'rows_per_partition': 'NONE'}
    AND comment = ''
    AND compaction = {'class': 'org.apache.cassandra.db.compaction.SizeTieredC
ompactionStrategy', 'max_threshold': '32', 'min_threshold': '4'}
    AND compression = {'chunk_length_in_kb': '64', 'class': 'org.apache.cassan
dra.io.compress.LZ4Compressor'}
    AND crc_check_chance = 1.0
    AND dclocal_read_repair_chance = 0.1
    AND default_time_to_live = 0
    AND gc_grace_seconds = 864000
    AND max_index_interval = 2048
    AND memtable_flush_period_in_ms = 0
    AND min_index_interval = 128
    AND read_repair_chance = 0.0
    AND speculative_retry = '99PERCENTILE';
```

There were around 400 million records before creating the following index and compaction started normally:

```
CREATE CUSTOM INDEX idx1 ON ds1.waypoints_subset (idx1col) USING 'com.stratio.
cassandra.lucene.Index' WITH OPTIONS = {'refresh_seconds': '60', 'schema': '{fields :{
          driverid : {type : "integer"}
         ,speed : {type : "double"}
         ,odometer : {type : "double"}
         }
        }'};
```

But I got the following exception in system.log :
```

ERROR [CompactionExecutor:14] 2016-05-27 10:45:29,221 CassandraDaemon.java:195 - Exception in thread Thread[CompactionExecutor:
14,1,main]
java.lang.IllegalStateException: Composite column is already fully constructed
        at org.apache.cassandra.db.marshal.CompositeType$Builder.add(CompositeType.java:434) ~[apache-cassandra-3.0.5.jar:3.0.5]
        at com.stratio.cassandra.lucene.key.KeyMapper.byteBuffer(KeyMapper.java:177) ~[cassandra-lucene-index-plugin-3.0.5.2-RC1-SNAPSHOT.jar:na]
        at com.stratio.cassandra.lucene.key.KeyMapper.byteBuffer(KeyMapper.java:164) ~[cassandra-lucene-index-plugin-3.0.5.2-RC1-SNAPSHOT.jar:na]
        at com.stratio.cassandra.lucene.key.KeyMapper.addFields(KeyMapper.java:191) ~[cassandra-lucene-index-plugin-3.0.5.2-RC1-SNAPSHOT.jar:na]
        at com.stratio.cassandra.lucene.IndexServiceWide.document(IndexServiceWide.java:116) ~[cassandra-lucene-index-plugin-3.0.5.2-RC1-SNAPSHOT.jar:na]
        at com.stratio.cassandra.lucene.IndexService.lambda$upsert$4(IndexService.java:389) ~[cassandra-lucene-index-plugin-3.0.5.2-RC1-SNAPSHOT.jar:na]
        at com.stratio.cassandra.lucene.util.TaskQueue.submitAsynchronous(TaskQueue.java:78) ~[cassandra-lucene-index-plugin-3.0.5.2-RC1-SNAPSHOT.jar:na]
        at com.stratio.cassandra.lucene.IndexService.upsert(IndexService.java:389) ~[cassandra-lucene-index-plugin-3.0.5.2-RC1-SNAPSHOT.jar:na]
        at com.stratio.cassandra.lucene.IndexWriterWide.lambda$null$1(IndexWriterWide.java:93) ~[cassandra-lucene-index-plugin-3.0.5.2-RC1-SNAPSHOT.jar:na]
        at java.util.Optional.ifPresent(Optional.java:159) ~[na:1.8.0_91]
        at com.stratio.cassandra.lucene.IndexWriterWide.lambda$finish$2(IndexWriterWide.java:91) ~[cassandra-lucene-index-plugin-3.0.5.2-RC1-SNAPSHOT.jar:na]
        at java.util.LinkedHashMap.forEach(LinkedHashMap.java:676) ~[na:1.8.0_91]
        at com.stratio.cassandra.lucene.IndexWriterWide.finish(IndexWriterWide.java:91) ~[cassandra-lucene-index-plugin-3.0.5.2-RC1-SNAPSHOT.jar:na]
        at java.lang.Iterable.forEach(Iterable.java:75) ~[na:1.8.0_91]
        at org.apache.cassandra.index.SecondaryIndexManager.indexPartition(SecondaryIndexManager.java:548) ~[apache-cassandra-3.0.5.jar:3.0.5]
        at org.apache.cassandra.db.Keyspace.indexPartition(Keyspace.java:538) ~[apache-cassandra-3.0.5.jar:3.0.5]
        at org.apache.cassandra.index.SecondaryIndexBuilder.build(SecondaryIndexBuilder.java:69) ~[apache-cassandra-3.0.5.jar:3.0.5]
        at org.apache.cassandra.db.compaction.CompactionManager$11.run(CompactionManager.java:1345) ~[apache-cassandra-3.0.5.jar:3.0.5]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_91]
        at java.util.concurrent.FutureTask.run(FutureTask.java:266) ~[na:1.8.0_91]
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142) ~[na:1.8.0_91]
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617) [na:1.8.0_91]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_91]
```

and following exception, when run a query by using the index (e.g. select * from ds1.waypoints_subset where idx1col='{ filter: {type:"range", field: "speed" , upper:10} }' limit 10;)

```
WARN  [SharedPool-Worker-2] 2016-05-27 10:53:45,884 AbstractLocalAwareExecutorService.java:169 - Uncaught exception on thread T
hread[SharedPool-Worker-2,5,main]: {}
java.lang.RuntimeException: org.apache.cassandra.index.IndexNotAvailableException: The secondary index 'idx1' is not yet availa
ble
        at org.apache.cassandra.service.StorageProxy$DroppableRunnable.run(StorageProxy.java:2449) ~[apache-cassandra-3.0.5.jar
:3.0.5]
        at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511) ~[na:1.8.0_91]
        at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$FutureTask.run(AbstractLocalAwareExecutorService.j
ava:164) ~[apache-cassandra-3.0.5.jar:3.0.5]
        at org.apache.cassandra.concurrent.AbstractLocalAwareExecutorService$LocalSessionFutureTask.run(AbstractLocalAwareExecu
torService.java:136) [apache-cassandra-3.0.5.jar:3.0.5]
        at org.apache.cassandra.concurrent.SEPWorker.run(SEPWorker.java:105) [apache-cassandra-3.0.5.jar:3.0.5]
        at java.lang.Thread.run(Thread.java:745) [na:1.8.0_91]
Caused by: org.apache.cassandra.index.IndexNotAvailableException: The secondary index 'idx1' is not yet available
        at org.apache.cassandra.db.ReadCommand.executeLocally(ReadCommand.java:356) ~[apache-cassandra-3.0.5.jar:3.0.5]
        at org.apache.cassandra.service.StorageProxy$LocalReadRunnable.runMayThrow(StorageProxy.java:1790) ~[apache-cassandra-3
.0.5.jar:3.0.5]
        at org.apache.cassandra.service.StorageProxy$DroppableRunnable.run(StorageProxy.java:2445) ~[apache-cassandra-3.0.5.jar
:3.0.5]
        ... 5 common frames omitted
```

which is fair enough. Appreciate any help.

**Comment by adelapena** [comment]
Hi @rpazuki,

Regarding the first error, this is a bug, [we have just fixed it](https://github.com/Stratio/cassandra-lucene-index/commit/4bc32536c034ac5970674aa486f1c7e7d09dd3ec). To apply the fix, drop your index, update the index plugin, and create it again.

Regarding the second error, this is just Cassandra telling you that [the index is not yet available because it's still being built](https://issues.apache.org/jira/browse/CASSANDRA-8505). Indexing 400 million records is going to take a while. Indexing speed depends on your hardware and data, but you can expect something around a few thousand rows per second per node. So, index building can take hours, depending on the number of nodes and replication factor. You can [turn on asynchronous indexing](https://github.com/Stratio/cassandra-lucene-index/blob/branch-3.0.5/doc/documentation.rst#indexing) to improve indexing speed.

**Comment by rpazuki** [comment]
Thanks a million @adelapena 

Cheers

**Comment by adelapena** [comment]
The fix is now available in [3.0.5.2](https://github.com/Stratio/cassandra-lucene-index/tree/3.0.5.2), [3.5.2](https://github.com/Stratio/cassandra-lucene-index/tree/3.5.2) and [3.0.6.0](https://github.com/Stratio/cassandra-lucene-index/tree/3.0.6.0) releases.
